substrate/frame/identity/src/lib.rs

```diff
@@ -925,7 +1179,99 @@ impl<T: Config> Pallet<T> {
 		fields: <T::IdentityInformation as IdentityInformationProvider>::FieldsIdentifier,
 	) -> bool {
 		IdentityOf::<T>::get(who)
-			.map_or(false, |registration| (registration.info.has_identity(fields)))
+			.map_or(false, |(registration, _username)| (registration.info.has_identity(fields)))
+	}
+
+	/// Calculate the deposit required for an identity.
+	fn calculate_identity_deposit(info: &T::IdentityInformation) -> BalanceOf<T> {
+		let bytes = info.encoded_size() as u32;
+		let byte_deposit = T::ByteDeposit::get().saturating_mul(<BalanceOf<T>>::from(bytes));
+		T::BasicDeposit::get().saturating_add(byte_deposit)
+	}
+
+	/// Validate that a username conforms to allowed characters/format. The function will validate
+	/// the characters in `username` and that `length` (if `Some`) conforms to the limit. It is not
+	/// expected to pass a fully formatted username here (i.e. one with any protocol-added
+	/// characters included, such as a `.`).
+	fn validate_username(username: &Vec<u8>, length: Option<u32>) -> DispatchResult {
+		// Verify input length before allocating a Vec with the user's input. `<` instead of `<=`
+		// because it needs one element for the point (`username` + `.` + `suffix`).
+		if let Some(l) = length {
+			ensure!(l < USERNAME_MAX_LENGTH, Error::<T>::InvalidUsername);
+		}
+		ensure!(
+			username.iter().all(|byte| byte.is_ascii_alphanumeric()),
```

Comment:
Not sure if we should maybe normalize the usernames to lower case, just like URLs. Otherwise `bob.dot` and `Bob.dot` would be different.